src/test/java/org/mockitousage/matchers/MatchersTest.java

```diff
@@ -624,4 +626,23 @@ public void nullable_matcher() throws Exception {
 
         verify(mock, times(2)).oneArg(nullable(Character.class));
     }
+
+    @Test
+    public void assertArg_matcher() throws Exception {
+        mock.oneArg("hello");
+
+        verify(mock).oneArg(assertArg((String it) -> assertEquals("hello", it)));
+    }
+
+    @Test
+    public void assertArg_matcher_fails_when_assertion_fails() throws Exception {
```

Comment:
I just realized that throwing an exception might cause stateful problems in our tracking mechanism. E.g. if we match on arguments and we throw an exception somewhere, I am not sure how Mockito recovers. Given that this PR passes the checks, I think this is fine, but let's codify this explicitly.

Can we add the following test cases?

```java
    @Test
    public void can_invoke_method_on_mock_after_assert_arg() throws Exception {
        mock.oneArg("hello");

        try {
            verify(mock).oneArg(assertArg((String it) -> assertEquals("not-hello", it)));
            fail("Should throw an exception");
        } catch (ComparisonFailure e) {
            // do nothing
        }

        mock.oneArg("hello");
    }

    @Test
    public void can_verify_on_mock_after_assert_arg() throws Exception {
        mock.oneArg("hello");

        try {
            verify(mock).oneArg(assertArg((String it) -> assertEquals("not-hello", it)));
            fail("Should throw an exception");
        } catch (ComparisonFailure e) {
            // do nothing
        }

        verify(mock).oneArg("hello");
    }
```

Reply:
@TimvdLippe tests added